File: frontend/src/lib/api/observability.ts
import { apiRequest } from "@/lib/api/request";

export type ObservabilityRange = {
  from: string;
  to: string;
};

export type ApiMetrics = {
  total_requests: number;
  failed_requests: number;
  error_rate: number | null;
  avg_latency_ms: number | null;
  p95_latency_ms: number | null;
  telemetry_missing: boolean;
};

export type LlmModelSummary = {
  model_name: string;
  event_count: number;
  error_count: number;
};

export type LlmMetrics = {
  total_events: number;
  failed_events: number;
  error_rate: number | null;
  avg_latency_ms: number | null;
  top_models: LlmModelSummary[];
  telemetry_missing: boolean;
};

export type IndexingMetrics = {
  total_jobs: number;
  succeeded_jobs: number;
  failed_jobs: number;
  in_progress_jobs: number;
  success_rate: number | null;
  telemetry_missing: boolean;
};

export type StorageMetrics = {
  total_documents: number;
  indexed_documents: number;
  failed_documents: number;
  pending_documents: number;
  total_chunks: number;
};

export type ObservabilitySnapshot = {
  organization_id: string;
  range: ObservabilityRange;
  generated_at: string;
  api_metrics: ApiMetrics;
  llm_metrics: LlmMetrics;
  indexing_metrics: IndexingMetrics;
  storage_metrics: StorageMetrics;
};

export type ObservabilityQuery = {
  from?: string;
  to?: string;
};

export async function getObservabilitySnapshot(
  query?: ObservabilityQuery,
): Promise<ObservabilitySnapshot> {
  const params = new URLSearchParams();
  if (query?.from) params.set("from", query.from);
  if (query?.to) params.set("to", query.to);
  const qs = params.toString();
  return apiRequest<ObservabilitySnapshot>(
    `/admin/observability${qs ? `?${qs}` : ""}`,
  );
}

File: backend/app/domains/admin/schemas/observability.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class ObservabilityRange(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    from_date: date = Field(alias="from")
    to_date: date = Field(alias="to")


class ApiMetrics(BaseModel):
    total_requests: int
    failed_requests: int
    error_rate: float | None
    avg_latency_ms: float | None
    p95_latency_ms: float | None
    telemetry_missing: bool


class LlmModelSummary(BaseModel):
    model_name: str
    event_count: int
    error_count: int


class LlmMetrics(BaseModel):
    total_events: int
    failed_events: int
    error_rate: float | None
    avg_latency_ms: float | None
    top_models: list[LlmModelSummary]
    telemetry_missing: bool


class IndexingMetrics(BaseModel):
    total_jobs: int
    succeeded_jobs: int
    failed_jobs: int
    in_progress_jobs: int
    success_rate: float | None
    telemetry_missing: bool


class StorageMetrics(BaseModel):
    total_documents: int
    indexed_documents: int
    failed_documents: int
    pending_documents: int
    total_chunks: int


ObservabilityStatus = Literal["healthy", "degraded", "unknown"]


class ObservabilitySnapshot(BaseModel):
    organization_id: str
    range: ObservabilityRange
    generated_at: datetime
    api_metrics: ApiMetrics
    llm_metrics: LlmMetrics
    indexing_metrics: IndexingMetrics
    storage_metrics: StorageMetrics
